# Ingest data from GitHub

In [0]:
%sh
rm -r /dbfs/ml_lab
mkdir /dbfs/ml_lab
wget -O /dbfs/ml_lab/customer1.csv https://raw.githubusercontent.com/sachin365123/CSV-files-for-Data-Science-and-Machine-Learning/refs/heads/main/Telco-Customer-Churn%20with%20Registry%20Models.csv

--2025-08-01 14:52:42--  https://raw.githubusercontent.com/sachin365123/CSV-files-for-Data-Science-and-Machine-Learning/refs/heads/main/Telco-Customer-Churn%20with%20Registry%20Models.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 189150 (185K) [text/plain]
Saving to: ‘/dbfs/ml_lab/customer1.csv’

     0K .......... .......... .......... .......... .......... 27%  696K 0s
    50K .......... .......... .......... .......... .......... 54%  982K 0s
   100K .......... .......... .......... .......... .......... 81% 3.20M 0s
   150K .......... .......... .......... ....                 100% 5.55M=0.1s

2025-08-01 14:52:43 (1.25 MB/s) - ‘/dbfs/ml_lab/customer1.csv’ saved [189150/189150]



# Read dataset

In [0]:
%python
from pyspark.sql.types import *
from pyspark.sql.functions import *
   
data = spark.read.format("csv").option("header", "true").load("dbfs:/ml_lab/customer1.csv")

# Display columns

In [0]:
data.columns

['SeniorCitizen',
 'PropertyValue',
 'tenure',
 'PhoneService',
 'MonthlyCharges',
 'TotalCharges']

# clean dataset: Remove Null values

In [0]:
data = data.dropna().select(col("PhoneService").astype("string"),
                          col("MonthlyCharges").astype("float"),
                          col("TotalCharges").astype("float"),
                          col("PropertyValue").astype("float"),
                          col("SeniorCitizen").astype("int"),
                          col("tenure").astype("int")
                          )
display(data.sample(0.2))

PhoneService MonthlyCharges TotalCharges PropertyValue SeniorCitizen tenure Yes 56.95 1889.5 3.0 0 34 No 42.3 1840.75 5.0 0 45 Yes 70.7 151.65 6.0 0 2 Yes 18.95 326.8 13.0 0 16 Yes 55.2 528.35 20.0 0 10 Yes 19.8 202.25 23.0 0 12 Yes 59.6 2970.3 26.0 0 49 Yes 55.3 1530.6 27.0 0 30 Yes 64.7 1093.1 31.0 0 17 Yes 95.5 181.65 33.0 1 2 Yes 106.35 3549.25 40.0 0 34 Yes 20.75 418.25 44.0 0 17 Yes 76.2 981.45 46.0 0 13 Yes 80.65 144.15 49.0 0 2 Yes 69.5 1752.65 54.0 1 25 Yes 24.95 894.3 60.0 0 34 Yes 107.5 7853.7 61.0 0 72 Yes 100.5 4707.1 62.0 0 47 Yes 54.4 957.1 65.0 0 18 Yes 94.4 857.25 66.0 0 9 Yes 78.9 3650.35 68.0 0 47 Yes 64.5 177.4 76.0 0 3 Yes 103.8 4327.5 85.0 0 42 Yes 62.15 2215.45 88.0 0 35 No 33.75 1009.25 91.0 0 29 Yes 100.9 7459.05 96.0 0 72 Yes 21.05 113.85 99.0 0 5 No 30.15 927.65 109.0 0 32 Yes 19.4 1363.25 111.0 0 72 Yes 96.75 5238.9 112.0 0 55 Yes 57.95 3042.25 113.0 0 52 Yes 20.65 835.15 120.0 0 41 Yes 20.7 112.75 123.0 0 6 Yes 19.75 19.75 130.0 0 1 No 58.6 1502.65 133.0 0 25 Yes 75.0 1778.5 138.0 0 23 Yes 84.3 5289.05 145.0 0 64 Yes 99.05 6416.7 147.0 0 65 Yes 64.5 1929.95 150.0 0 30 Yes 84.15 1821.95 158.0 0 22 Yes 20.15 419.9 159.0 0 22 Yes 25.7 251.6 161.0 0 10 Yes 82.4 1592.35 163.0 0 20 No 45.3 2651.2 170.0 1 58 Yes 104.4 242.8 173.0 0 2 Yes 81.95 1181.75 174.0 0 14 Yes 24.7 780.2 177.0 0 34 Yes 112.25 8041.65 200.0 0 72 Yes 68.85 2958.95 206.0 0 44 Yes 79.9 79.9 209.0 0 1 No 56.6 3789.2 219.0 0 66 No 50.6 1288.75 227.0 0 24 Yes 81.15 2545.75 228.0 0 31 Yes 108.0 5760.65 232.0 0 54 Yes 54.4 114.1 237.0 0 2 No 29.6 299.05 238.0 0 10 Yes 73.15 1305.95 239.0 0 18 Yes 54.8 731.3 246.0 1 13 Yes 113.3 5032.25 266.0 0 45 Yes 88.1 5526.75 267.0 0 61 Yes 24.9 1195.25 268.0 0 50 Yes 19.35 1099.6 270.0 0 59 Yes 70.6 70.6 276.0 0 1 Yes 64.95 1339.8 279.0 1 21 No 55.0 771.95 280.0 0 14 Yes 50.55 244.75 281.0 0 5 Yes 55.15 322.9 282.0 0 6 Yes 51.2 498.25 283.0 0 10 Yes 76.0 1783.6 287.0 0 22 Yes 44.35 927.15 288.0 0 20 Yes 70.0 70.0 289.0 0 1 Yes 61.2 2021.2 293.0 0 35 Yes 20.25 20.25 297.0 0 1 Yes 64.9 1509.8 302.0 0 23 Yes 74.9 2303.35 308.0 1 32 Yes 19.25 19.25 312.0 0 1 Yes 89.05 1886.25 313.0 0 22 Yes 48.6 422.3 325.0 0 9 Yes 70.35 139.05 329.0 1 2 No 35.9 973.65 330.0 1 28 Yes 88.15 6148.45 339.0 0 68 Yes 94.8 3565.65 340.0 0 36 Yes 20.5 1328.15 345.0 0 63 Yes 19.45 1303.5 349.0 0 70 No 44.95 996.85 350.0 0 23 Yes 54.1 889.0 358.0 1 16 Yes 104.95 2566.5 379.0 0 25 Yes 50.65 50.65 383.0 0 1 Yes 44.0 44.0 391.0 0 1 Yes 84.8 2309.55 393.0 1 27 Yes 24.45 1066.15 413.0 0 46 Yes 74.95 2149.05 414.0 0 29 Yes 70.65 3545.05 416.0 1 48 Yes 98.7 6858.9 418.0 0 70 Yes 24.8 1024.7 419.0 0 40 Yes 25.7 1110.5 425.0 0 44 Yes 51.65 2838.55 427.0 1 55 Yes 89.85 3161.2 440.0 0 35 Yes 112.95 1882.55 442.0 0 16 Yes 64.8 1166.7 445.0 0 18 Yes 94.75 759.55 450.0 0 8 Yes 81.05 2227.1 455.0 0 28 Yes 20.65 330.6 472.0 1 17 Yes 76.05 2747.2 478.0 1 35 Yes 19.65 225.75 482.0 0 11 Yes 88.6 6201.95 489.0 0 72 No 28.45 131.05 499.0 1 5 Yes 99.7 1238.45 500.0 0 12 Yes 79.45 5502.55 505.0 0 69 Yes 19.85 19.85 510.0 0 1 Yes 94.3 1818.3 511.0 1 20 Yes 75.6 1758.6 524.0 0 23 Yes 91.25 4738.3 526.0 0 52 Yes 94.7 5430.35 528.0 0 58 Yes 21.1 490.65 531.0 0 24 Yes 89.55 1021.75 536.0 0 12 No 45.55 597.0 545.0 0 13 Yes 101.9 5265.5 549.0 1 50 Yes 115.6 8220.4 552.0 0 72 Yes 68.75 3808.0 556.0 0 56 Yes 107.25 6033.3 561.0 0 56 Yes 70.5 597.9 578.0 0 9 Yes 94.55 3365.4 579.0 1 35 Yes 19.65 38.8 580.0 0 2 Yes 89.2 346.2 583.0 1 4 Yes 111.9 8071.05 591.0 0 72 Yes 84.3 2438.6 594.0 0 30 Yes 19.95 47.7 602.0 0 2 Yes 109.55 3608.0 605.0 0 32 Yes 105.25 6786.4 610.0 1 65 Yes 69.5 3418.2 615.0 0 50 Yes 92.05 5755.8 622.0 0 62 Yes 20.0 137.6 625.0 0 7 Yes 89.85 6562.9 633.0 0 72 No 54.55 2236.2 635.0 0 40 Yes 70.9 4911.35 638.0 0 68 Yes 104.55 5794.65 639.0 1 56 Yes 85.25 855.3 640.0 0 10 Yes 89.55 89.55 643.0 0 1 Yes 25.25 394.85 645.0 0 15 Yes 69.65 3442.15 648.0 0 50 Yes 89.5 161.5 649.0 1 2 Yes 69.55 222.3 651.0 0 3 Yes 24.8 475.25 654.0 0 17 Yes 19.85 35.9 659.0

# split my Data into Train and Test

In [0]:
splits = data.randomSplit([0.7, 0.3])
train = splits[0]
test = splits[1]
print ("Training Rows:", train.count(), " Testing Rows:", test.count())

Training Rows: 4913  Testing Rows: 2119


# Segregate into Category and numerical columns

In [0]:
from pyspark.ml import Pipeline
from pyspark.ml.feature import StringIndexer, VectorAssembler, MinMaxScaler
from pyspark.ml.classification import LogisticRegression
   
catFeature = "PhoneService"
numFeatures = ['MonthlyCharges', 'TotalCharges', 'PropertyValue',  'tenure']

# Define the feature engineering and model training algorithm steps

In [0]:
catIndexer = StringIndexer(inputCol=catFeature, outputCol=catFeature + "Idx")
numVector = VectorAssembler(inputCols=numFeatures, outputCol="numericFeatures")
numScaler = MinMaxScaler(inputCol = numVector.getOutputCol(), outputCol="normalizedFeatures")
featureVector = VectorAssembler(inputCols=["PhoneServiceIdx", "normalizedFeatures"], outputCol="Features")


#### using the StringIndexer from PySpark to convert a categorical feature into a numerical index. 

#### inputCol=catFeature: This is the name of the categorical column you want to index.

#### outputCol=catFeature + "Idx": The name of the new column that will hold the indexed values (integers representing each category).

🧠 Example:
#### If your DataFrame looks like:

#### tenure	MonthlyCharges	TotalCharges
#### 1	      29.85	          29.85
#### 34	    56.95	          1889.50

#### Then after applying VectorAssembler, the output will be:

| numericFeatures        |
| ---------------------- |
| \[1.0, 29.85, 29.85]   |
| \[34.0, 56.95, 1889.5] |


### Combines multiple numerical columns into a single vector column called "numericFeatures"
#### Scales the "numericFeatures" vector to a 0–1 range, producing "normalizedFeatures"
#### Combines both categorical (now indexed) and scaled numerical features into one feature vector "Features" to be used in the model.


# Logistic Regression

In [0]:
algo = LogisticRegression(labelCol="SeniorCitizen", featuresCol="Features", maxIter=10, regParam=0.3)

StringIndexer_bfa8c6cf5ad1

# Chain the steps as stages in a pipeline

In [0]:
pipeline = Pipeline(stages=[catIndexer, numVector, numScaler, featureVector, algo])

#### You’re creating a sequence of transformations and a model training step, combined into a single pipeline. Here's what each stage likely represents:

#### catIndexer:
StringIndexer to convert a categorical column to an index.

#### numVector:
VectorAssembler to merge numeric columns into a single vector.

#### numScaler (optional but common):
Something like MinMaxScaler or StandardScaler to normalize the numeric features.

#### featureVector:
Another VectorAssembler that likely combines:

#### The scaled numeric vector

The indexed categorical feature(s) into a final features vector.

#### algo:
The ML model you’re training, like:

#### LogisticRegression()



# Use the pipeline to prepare data and fit the model algorithm

In [0]:
model = pipeline.fit(train)
print ("Model trained!")

Uploading artifacts:   0%|          | 0/4 [00:00<?, ?it/s]

Model trained!


In [0]:
prediction = model.transform(test)
predicted = prediction.select("Features", "probability", col("prediction").astype("Int"), col("SeniorCitizen").alias("trueLabel"))
display(predicted)

Features probability prediction trueLabel Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06069652120865399, 0.015259457439860666, 0.7058907228229938, 0.08450704225352113)) Map(vectorType -> dense, length -> 2, values -> List(0.8693606526012754, 0.13063934739872463)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06119402605502759, 6.47125808714733E-4, 0.46272054638588506, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.8699397317711933, 0.13006026822880667)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06119402605502759, 6.47125808714733E-4, 0.6590779738190097, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.8694173549622236, 0.1305826450377764)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06119402605502759, 6.47125808714733E-4, 0.9285714285714286, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.8686975099834627, 0.13130249001653727)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.061691549879994555, 6.529038495733589E-4, 0.5334376778599886, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.869716109618651, 0.13028389038134902)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06218905472636816, 0.003189405818753267, 0.386881047239613, 0.014084507042253521)) Map(vectorType -> dense, length -> 2, values -> List(0.8700726576286649, 0.12992734237133508)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06318408339770872, 0.01979511570090298, 0.1730221969265794, 0.11267605633802817)) Map(vectorType -> dense, length -> 2, values -> List(0.8706176180333846, 0.12938238196661545)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06517412176179649, 6.933490335351306E-4, 0.02618099032441662, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.870811423262365, 0.12918857673763495)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06517412176179649, 6.933490335351306E-4, 0.03856004553215709, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.8707787310053043, 0.12922126899469566)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06616915043313705, 7.049048948426605E-4, 0.8524473534433694, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.8685418800496828, 0.1314581199503172)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06616915043313705, 0.0035707476990337003, 0.8712293682413205, 0.014084507042253521)) Map(vectorType -> dense, length -> 2, values -> List(0.868485463831157, 0.131514536168843)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06766168395085122, 7.222385765990945E-4, 0.9247296528173022, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.8682399503757062, 0.13176004962429377)) 0 1 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06766168395085122, 0.019777780917097934, 0.5714285714285714, 0.08450704225352113)) Map(vectorType -> dense, length -> 2, values -> List(0.8690999415957767, 0.13090005840422325)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06815920777581817, 0.007049048948426605, 0.40338645418326696, 0.028169014084507043)) Map(vectorType -> dense, length -> 2, values -> List(0.8695701637952787, 0.13042983620472126)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06815920777581817, 0.12461215796150968, 0.6613545816733069, 0.5774647887323944)) Map(vectorType -> dense, length -> 2, values -> List(0.8684905871002518, 0.13150941289974816)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06865671262219178, 7.337944379066243E-4, 0.18867387592487195, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.8701327177527389, 0.12986728224726107)) 0 0 Map(vectorType -> dense, length -> 5, values -> List(1.0, 0.06865671262219178, 7.337944379066243E-4, 0.3256972111553785, 0.0)) Map(vectorType -> dense, length -> 2, values -> List(0.8697688369793174, 0.1302311630206826)) 0 0 Map(vecto

# Generate evaluation metrics

In [0]:

from pyspark.ml.evaluation import MulticlassClassificationEvaluator
   
evaluator = MulticlassClassificationEvaluator(labelCol="SeniorCitizen", predictionCol="prediction")

# Simple accuracy

In [0]:
accuracy = evaluator.evaluate(prediction, {evaluator.metricName:"accuracy"})
print("Accuracy:", accuracy)
   

Accuracy: 0.8381311939594148


# Class metrics

In [0]:

labels = [0, 1]
print("\nIndividual class metrics:")
for label in sorted(labels):
    print ("Class %s" % (label))
   
    # Precision
    precision = evaluator.evaluate(prediction, {evaluator.metricLabel:label,
                                                    evaluator.metricName:"precisionByLabel"})
    print("\tPrecision:", precision)
   
    # Recall
    recall = evaluator.evaluate(prediction, {evaluator.metricLabel:label,
                                             evaluator.metricName:"recallByLabel"})
    print("\tRecall:", recall)
   
    # F1 score
    f1 = evaluator.evaluate(prediction, {evaluator.metricLabel:label,
                                         evaluator.metricName:"fMeasureByLabel"})
    print("\tF1 Score:", f1)
   



Individual class metrics:
Class 0
	Precision: 0.8381311939594148
	Recall: 1.0
	F1 Score: 0.9119383825417201
Class 1
	Precision: 0.0
	Recall: 0.0
	F1 Score: 0.0


# Weighed (overall) metrics

In [0]:
overallPrecision = evaluator.evaluate(prediction, {evaluator.metricName:"weightedPrecision"})
print("Overall Precision:", overallPrecision)
overallRecall = evaluator.evaluate(prediction, {evaluator.metricName:"weightedRecall"})
print("Overall Recall:", overallRecall)
overallF1 = evaluator.evaluate(prediction, {evaluator.metricName:"weightedFMeasure"})
print("Overall F1 Score:", overallF1) 

Overall Precision: 0.7024638982878342
Overall Recall: 0.8381311939594148
Overall F1 Score: 0.7643240053771094


# now stop your running Cluster

#### --------------------------------------------------------------

# Select Experiments on the left pane.

####  Select the experiment generated with your notebook’s name and view the details page for the most recent experiment run.

####   Use the Register Model button to register the model that was logged in that experiment and when prompted, create a new model named Penguin Predictor.

####  When the model has been registered, view the Models page (in the navigation bar on the left) and select the Penguin Predictor model.
####  In the page for the Penguin Predictor model, use the Use model for inference button to create a new real-time endpoint with the following settings:
####  Model: Penguin Predictor
####  Model version: 1
####  Endpoint: predict-penguin
####  Compute size: Small
####  The serving endpoint is hosted in a new cluster, which it may take several minutes to create.


####   When the endpoint has been created, use the Query endpoint button at the top right to open an interface from which you can test the endpoint. 

#### Then in the test interface, on the Browser tab, enter the following JSON request and use the Send Request button to call the endpoint and generate a prediction.


##### 
#### Paste the sample data into "Serving" from left pane -> Serving endpoints -> Browser -> Request --> Send Request

## Sample Data

In [0]:

{"dataframe_records": [
        {   "PhoneService": "No",
            "MonthlyCharges": 29.85,
            "TotalCharges": 29.85,
            "PropertyValue": 2,
            "tenure": 1
        } 
    ]
}

### code copied from Query Endpoint option, once Register model is reday for "Use model for inference"


# Copy token from Datbricks Setting option 
## Setting -> User -> Developer -> Access tokens -> Manage -> Generate New Token

In [0]:

import os

# Replace with your actual Databricks token
os.environ["DATABRICKS_TOKEN"] = "dapi3d4cacdba3bea15fccaafb9f4eb6fde1"

In [0]:
import os
import requests
import numpy as np
import pandas as pd
import json

def create_tf_serving_json(data):
    return {'inputs': {name: data[name].tolist() for name in data.keys()} if isinstance(data, dict) else data.tolist()}

def score_model(dataset):
    url = 'https://adb-2596735487416449.9.azuredatabricks.net/serving-endpoints/End_Point_DevOps_Aug_01/invocations'
    headers = {'Authorization': f'Bearer {os.environ.get("DATABRICKS_TOKEN")}', 'Content-Type': 'application/json'}
    ds_dict = {'dataframe_split': dataset.to_dict(orient='split')} if isinstance(dataset, pd.DataFrame) else create_tf_serving_json(dataset)
    data_json = json.dumps(ds_dict, allow_nan=True)
    response = requests.request(method='POST', headers=headers, url=url, data=data_json)
    if response.status_code != 200:
        raise Exception(f'Request failed with status {response.status_code}, {response.text}')
    return response.json()

# Test above end points for sample input

In [0]:
import pandas as pd

# Define the input record
input_data = {
    "dataframe_records": [
        {
            "PhoneService": "No",
            "MonthlyCharges": 29.85,
            "TotalCharges": 29.85,
            "PropertyValue": 2,
            "tenure": 1
        }
    ]
}

# For this test data output is SeniorCitizen: 0

# Convert it to a DataFrame
df_input = pd.DataFrame(input_data["dataframe_records"])

display(df_input)

PhoneService MonthlyCharges TotalCharges PropertyValue tenure No 29.85 29.85 2 1

# Call the score_model function

In [0]:
# Call the function
response = score_model(df_input)

# View the prediction result
print(response)

{'predictions': [0.0]}


# Delete the endpoint
## When the endpoint is not longer required, you should delete it to avoid unnecessary costs.

#### In the predict-penguin endpoint page, in the ⁝ menu, select Delete.